### PR TITLE
[inductor] Memory planning

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112178

This was originally @jansel's PR:
https://github.com/pytorch/pytorch/pull/102625, which I've built upon.

This diff implements static memory planning. It's disabled by default
while we examine its performance.

We use a greedy-by-size approach. For dynamic shapes, the sizes of the
example inputs are used as estimates when making planning decisions. We
generate expressions to calculate the actual memory offsets and sizes at
runtime when the values of the dynamic shapes are known. In order to
simplify these calculations, we have organized the allocations into a
tree that branches on space (address offsets) and time (live ranges).
Finally, we need to align these offsets, so we have added an `align`
sympy Expr to express these calculations.

Some limitations:

1. It is only enabled during inference for now. Enabling it for training
   increases peak memory usage as we allocate all the memory needed for
   training upfront, before freeing the memory allocated during
   inference. We can probably address this by doing planning for both
   the inference and training passes together.
2. It doesn't work with PyTorch Distributed, because kernels like
   AllGatherIntoTensor codegen strings which do memory operations. We
   can fix this down the line by having them emit MemoryPlanningLines
   instead.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: build_variables.bzl [hunk 1]
```diff
@@ -437,6 +437,7 @@ core_sources_full_mobile = core_sources_full_mobile_no_backend_interface + [
 ]
 
 core_sources_full = core_sources_full_mobile + [
+    "torch/csrc/inductor/inductor_ops.cpp",
```
Comment (desertfire) [inline build_variables.bzl:440]: This file is already included under `inductor_core_resources`.

File: third_party/fbgemm
Comment (desertfire) [inline third_party/fbgemm:1]: Please revert this

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -372,6 +372,10 @@ def _print_Min(self, expr):
             il = "{" + ", ".join(args) + "}"
             return f"std::min({il})"
 
+    def _print_align(self, expr):
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp.py:375]: Did you mean to override this?

**Review by desertfire — APPROVED** [review]

**Review by desertfire — COMMENTED** [review]
BTW, did you submit a benchmark run at https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml? I saw some failed runs from you, so wasn't sure the status.

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (int3) [reply torch/_inductor/codegen/cpp.py:375]: This isn't an override. We have defined `align` as a custom `sympy.Function`, so we need to teach the printer how to print this new sympy expression.

**Comment by int3** [comment]
Oh, I didn't realize submitting the PR would kick off a benchmark run. I thought this memory planning code would need to land first to run the benchmark successfully, so I was preparing the changes to the CI config as a separate PR.

Looks like the error message is

> you may only define up to 10 `inputs` for a `workflow_dispatch` event

Seems like there are too many benchmark options now...

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (desertfire) [reply torch/_inductor/codegen/cpp.py:375]: Oh, I thought you added `_print_align` to the base class in `common.py`.

Comment (int3) [reply torch/_inductor/codegen/cpp.py:375]: Ohh. I thought the base `ExprPrinter` class was in a sympy library; I see now that it is defined by us. I'll move the two definitions of `_print_align` into the base class.

File: torch/_inductor/codegen/wrapper.py [hunk 3]
```diff
@@ -473,10 +475,15 @@ def generate_end(self, result):
         return
 
     def generate_extern_kernel_alloc(self, extern_kernel, args):
+        ending = self.ending
+        if config.memory_planning and "view_as_complex" in str(extern_kernel.kernel):
+            # view operation fallbacks cause issues since inductor
+            # doesn't know the memory is still needed and might reuse it.
+            ending = f".clone(){ending}"
```
Comment (jansel) [inline torch/_inductor/codegen/wrapper.py:482]: cc @htyu 

we will need to remove this workaround once we have a lowering for `view_as_complex`

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/wrapper.py (hunk 3 above)
Comment (htyu) [reply torch/_inductor/codegen/wrapper.py:482]: We model the aliasing and memory reuse on `ComplexView` node (https://github.com/pytorch/pytorch/blob/c14c4efc0eb97ccc471ac6cbdcc659417227ae97/torch/_inductor/ir.py#L4446) . Is that sufficient?

Comment (jansel) [reply torch/_inductor/codegen/wrapper.py:482]: Yeah that should work, just need to handle aten.view_as_complex.

**Comment by int3** [comment]
Thanks all for the reviews 😄 

@desertfire I think you accidentally edited my comment when you meant to reply lol. But cool, lmk when you've fixed the CI and I will try it again

**Comment by desertfire** [comment]
It is tricky to get the import on Windows test working correctly, although we are even not able to run Inductor on Windows.

**Comment by int3** [comment]
Hm, why wasn't the test run as part of the land?

**Comment by huydhn** [comment]
> Hm, why wasn't the test run as part of the land?

ROCm runners are in a bit of a shortage atm, which cause long queue. So, we have decided to move ROCm jobs to a separate workflow (triggered by `ciflow/rocm`).  AMD is working to mitigate the issue.

**Pull request closed without merging**